File: src/data/seed.js
/**
 * Seed data for reconciliation demo. Deterministic pseudo-random values.
 */
const seedData = (() => {
  let seed = 123456;
  function rand() {
    seed = (seed * 9301 + 49297) % 233280;
    return seed / 233280;
  }

  const platformTransactions = [];
  const statuses = ['settled', 'captured'];
  const customerPrefixes = ['ACCT', 'CUST'];

  for (let i = 1; i <= 50; i += 1) {
    const id = `txn_${String(i).padStart(3, '0')}`;
    const day = String((i % 31) + 1).padStart(2, '0');
    let amount = Number((5 + rand() * 495).toFixed(2));
    if (id === 'txn_017') {
      amount = 142.87;
    }
    platformTransactions.push({
      transaction_id: id,
      customer_id: `${customerPrefixes[i % customerPrefixes.length]}_${String(i).padStart(3, '0')}`,
      amount,
      currency: 'GBP',
      created_at: `2025-03-${day}T0${(i % 24) + 1}:00:00Z`,
      status: i % 10 === 0 ? 'pending' : statuses[i % statuses.length],
    });
  }

  const bankSettlements = [];

  const tPlusN = (createdAt, n = 2) => {
    const d = new Date(createdAt);
    d.setUTCDate(d.getUTCDate() + n);
    return d.toISOString().slice(0, 10);
  };

  for (const tx of platformTransactions) {
    if (tx.transaction_id === 'txn_010') {
      continue; // unmatched
    }

    let settled_amount = tx.amount;
    let settlement_date = tPlusN(tx.created_at, 2);

    if (tx.transaction_id === 'txn_017') {
      settled_amount = 142.86;
    }
    if (tx.transaction_id === 'txn_031') {
      settlement_date = '2025-04-02';
    }

    bankSettlements.push({
      settlement_id: `stl_${tx.transaction_id}`,
      transaction_ref: tx.transaction_id,
      settled_amount,
      settlement_date,
      batch_id: `batch_${String((Number(tx.transaction_id.split('_')[1]) % 5) + 1).padStart(2, '0')}`,
    });

    if (tx.transaction_id === 'txn_022') {
      bankSettlements.push({
        settlement_id: `stl_${tx.transaction_id}_dup`,
        transaction_ref: tx.transaction_id,
        settled_amount,
        settlement_date,
        batch_id: `batch_${String((Number(tx.transaction_id.split('_')[1]) % 5) + 1).padStart(2, '0')}`,
      });
    }
  }

  bankSettlements.push({
    settlement_id: 'stl_orphan_001',
    transaction_ref: 'txn_orphan_refund',
    settled_amount: -67.5,
    settlement_date: '2025-03-28',
    batch_id: 'batch_09',
  });

  return { platformTransactions, bankSettlements };
})();

export const { platformTransactions, bankSettlements } = seedData;

File: backend/app/data/seed.py
from datetime import datetime, timedelta


def deterministic_random(seed):
    while True:
        seed = (seed * 9301 + 49297) % 233280
        yield seed / 233280


def generate_data():
    rand = deterministic_random(42)
    transactions = []
    payment_methods = ['card', 'bank_transfer', 'wallet']
    merchants = ['Acme Toys', 'Nimbus Coffee', 'Grove Books', 'Stellar Apparel', 'Quantum Goods']

    for i in range(1, 51):
        tid = f'txn_{i:03d}'
        amount = float(f'{5 + next(rand) * 495:.2f}')
        if tid == 'txn_017':
            amount = 142.87
        created_at = datetime(2025, 3, 1) + timedelta(days=(i - 1) % 31, hours=(i % 24))
        if tid == 'txn_031':
            created_at = datetime(2025, 3, 28, 12)
        transactions.append({
            'transaction_id': tid,
            'customer_id': f'CUST_{i:03d}',
            'amount': amount,
            'currency': 'GBP',
            'created_at': created_at.isoformat(),
            'status': 'settled' if i % 10 != 0 else 'pending',
            'merchant_name': merchants[i % len(merchants)],
            'payment_method': payment_methods[i % len(payment_methods)],
        })

    settlements = []
    for tx in transactions:
        if tx['transaction_id'] == 'txn_010':
            continue

        settlement_date = datetime.fromisoformat(tx['created_at']) + timedelta(days=2)
        if tx['transaction_id'] == 'txn_031':
            settlement_date = datetime(2025, 4, 2, 10)

        settled_amount = tx['amount']
        if tx['transaction_id'] == 'txn_017':
            settled_amount = 142.86

        sid = f'SETL_{tx["transaction_id"][4:]}'
        settlements.append({
            'settlement_id': sid,
            'transaction_ref': tx['transaction_id'],
            'settled_amount': settled_amount,
            'settlement_date': settlement_date.date().isoformat(),
            'batch_id': f'BATCH_{(int(tx["transaction_id"][4:]) % 10) + 1:03d}',
            'bank_reference': f'BR_{sid}',
        })

        if tx['transaction_id'] == 'txn_022':
            settlements.append({
                'settlement_id': 'SETL_089',
                'transaction_ref': 'txn_022',
                'settled_amount': settled_amount,
                'settlement_date': settlement_date.date().isoformat(),
                'batch_id': 'BATCH_004',
                'bank_reference': 'BR_SETL_089',
            })
            settlements.append({
                'settlement_id': 'SETL_090',
                'transaction_ref': 'txn_022',
                'settled_amount': settled_amount,
                'settlement_date': settlement_date.date().isoformat(),
                'batch_id': 'BATCH_004',
                'bank_reference': 'BR_SETL_090',
            })
            settlements = [s for s in settlements if s['settlement_id'] != sid]

    settlements.append({
        'settlement_id': 'SETL_REFUND',
        'transaction_ref': None,
        'settled_amount': -67.5,
        'settlement_date': '2025-03-28',
        'batch_id': 'BATCH_009',
        'bank_reference': 'BR_REFUND_001',
    })

    return transactions, settlements


transactions, settlements = generate_data()
